src/ol/tilegrid/tilegrid.js

```diff
@@ -427,9 +391,10 @@ ol.tilegrid.TileGrid.prototype.getTileCoordForXYAndResolution_ = function(
   var origin = this.getOrigin(z);
   var tileSize = ol.size.toSize(this.getTileSize(z), this.tmpSize_);
 
-  var adjust = reverseIntersectionPolicy ? 0.5 : 0;
-  var xFromOrigin = ((x - origin[0]) / resolution + adjust) | 0;
-  var yFromOrigin = ((y - origin[1]) / resolution + adjust) | 0;
+  var adjustX = reverseIntersectionPolicy ? 0.5 : 0;
+  var adjustY = reverseIntersectionPolicy ? 0 : 0.5;
+  var xFromOrigin = Math.floor((x - origin[0]) / resolution + adjustX);
+  var yFromOrigin = Math.floor((y - origin[1]) / resolution + adjustY);
```

Comment:
Two reasons:

1. The change from `(whatever) | 0` to `Math.floor(whatever)` was necessary because `(Infinity) | 0` is `0`, but should be `Infinity`.
1. The `y`adjustment for adjusting to screen pixels was previously incorrect, because screen coordinates increase downwards, but map coordinates increase upwards. The tests caught this problem after changing the origin to the top-left corner of the extent.